File: sdks/kotlin/client/src/commonMain/kotlin/com/algolia/client/configuration/internal/FlapjackAgent.kt
package com.flapjackhq.client.configuration.internal

import com.flapjackhq.client.configuration.AgentSegment

/** Handles to handle algolia agent segments. */
internal class FlapjackAgent(clientVersion: String) {

  private val segments = mutableSetOf(AgentSegment("Flapjack for Kotlin", clientVersion))

  fun add(segment: AgentSegment): Boolean = segments.add(segment)

  fun add(segments: List<AgentSegment>): Boolean = this.segments.addAll(segments)

  fun remove(segment: AgentSegment): Boolean = segments.remove(segment)

  override fun toString(): String = segments.joinToString("; ") { it.formatted() }

  private fun AgentSegment.formatted(): String = buildString {
    append(value)
    version?.let { version ->
      append(" (")
      append(version)
      append(")")
    }
  }
}

File: sdks/dart/packages/client_core/lib/src/transport/flapjack_agent.dart
import 'dart:core';

import 'package:flapjack_client_core/src/config/agent_segment.dart';

/// Handles Flapjack agent segments.
///
/// An instance of this class maintains a set of [AgentSegment]s, and provides
/// methods to add, remove, and format these segments.
final class FlapjackAgent {
  final Set<AgentSegment> _segments = {};

  /// Constructs an [FlapjackAgent] with the provided [clientVersion].
  FlapjackAgent(String clientVersion) {
    _segments.add(
      AgentSegment(value: "Flapjack for Dart", version: clientVersion),
    );
  }

  /// Adds a new [segment] to the agent segments.
  bool add(AgentSegment segment) => _segments.add(segment);

  /// Adds all [segments] to the agent segments.
  void addAll(Iterable<AgentSegment> segments) => _segments.addAll(segments);

  /// Removes [segment] from the agent segments.
  bool remove(AgentSegment segment) => _segments.remove(segment);

  /// Formats the agent segments into a semicolon-separated string.
  String formatted() => _segments.map((it) => it.formatted()).join("; ");

  @override
  String toString() {
    return 'FlapjackAgent{segments: $_segments}';
  }
}

/// Provides a formatted string representation for [AgentSegment].
extension on AgentSegment {
  String formatted() {
    StringBuffer sb = StringBuffer();
    sb.write(value);
    if (version != null) {
      sb.write(" (");
      sb.write(version);
      sb.write(")");
    }
    return sb.toString();
  }
}
